tests/test_box_Box.py

```diff
@@ -112,6 +112,38 @@ def test_wrap_multiple_images(self):
         points = np.array(points)
         npt.assert_allclose(box.wrap(points)[0, 0], -2, rtol=1e-6)
 
+    def test_out_default(self):
+        box = freud.box.Box(2, 2, 2, 1, 0, 0)
+        points = [[10, -5, -5], [0, 0.5, 0]]
+        npt.assert_allclose(box.wrap(points)[0, 0], -2, rtol=1e-6)
+
+    def test_out_provided_with_input_array(self):
+        box = freud.box.Box(2, 2, 2, 1, 0, 0)
+        points = [[10, -5, -5], [0, 0.5, 0]]
+        npt.assert_allclose(box.wrap(points, out=points)[0, 0], -2, rtol=1e-6)
+
+    def test_out_provided_with_new_array(self):
+        box = freud.box.Box(2, 2, 2, 1, 0, 0)
+        points = [[10, -5, -5], [0, 0.5, 0]]
+        points = np.array(points, dtype=np.float32)
+        new_array = np.zeros(points.shape, dtype=np.float32)
+        npt.assert_allclose(box.wrap(points, out=new_array)[0, 0], -2, rtol=1e-6)
+
+    def test_out_provided_with_array_with_wrong_shape(self):
+        box = freud.box.Box(2, 2, 2, 1, 0, 0)
+        points = [[10, -5, -5], [0, 0.5, 0]]
+        points = np.array(points, dtype=np.float32)
+        new_array = np.zeros((np.asarray(points.shape) - 1), dtype=np.float32)
+        with pytest.raises(ValueError):
+            npt.assert_allclose(box.wrap(points, out=new_array)[0, 0], -2, rtol=1e-6)
+
+    def test_out_provided_with_array_with_wrong_dtype(self):
+        box = freud.box.Box(2, 2, 2, 1, 0, 0)
+        points = [[10, -5, -5], [0, 0.5, 0]]
+        points = np.array(points, dtype=np.int)
+        new_array = np.zeros(points.shape, dtype=np.int64)
+        npt.assert_allclose(box.wrap(points, out=new_array)[0, 0], -2, rtol=1e-6)
```

Comment:
Should this fail?

Reply:
Yeah I think you're right. Actually I have a question about this: what happens here is that if the ```out``` array is not ```float32```, the returned array of ```box.wrap``` should give the right thing, but the ```out``` array wouldn't be overwritten? 
I'm not sure what exactly happens after ```out = freud.util._convert_array(out, shape=vecs.shape)```. It seems to me that even though a new array would be allocated, aren't we just using this new array for the rest of ```box.wrap```?

Reply:
I'm not immediately sure what is supposed to happen, we can come back to this after resolving the other comments.

Reply:
I added a type check in ```box.pyx``` just to make the test pass first.